File: packages/pricing-engine/src/engine.ts
import {
  DEMAND_EPSILON,
  RENORMALIZE_EPSILON,
  RENORMALIZE_MAX_ITERATIONS,
  RENORMALIZE_TOLERANCE,
  roundIntoRange,
} from '@bolsa/shared';
import type { ExplainEntry, ProductState, TickInput, TickOutput } from './types';
import { validateTickInput } from './validate';

function clip(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

function overrideAppliesAt(product: ProductState, tick: number): boolean {
  return product.manualOverride !== undefined && tick <= product.manualOverride.untilTick;
}

const NEUTRAL_EXPLAIN: ExplainEntry = { x: 0, r: 0, k: 0, delta: 0, c: 1 };

interface Working {
  product: ProductState;
  /** New smoothed demand after the EWMA update. */
  demand: number;
  /** Unrounded price this tick started from. */
  previous: number;
  /** Candidate price as a float; rounded to cents only at the very end. */
  candidate: number;
  explain: ExplainEntry;
  /** Pinned products keep their price: the group cannot move them. */
  pinned: boolean;
}

function previousRawPrice(product: ProductState): number {
  return clip(product.rawPrice ?? product.currentPrice, product.minPrice, product.maxPrice);
}

/**
 * Spec 5.3, one tick of the market.
 *
 * Pure and deterministic: the same input always produces the same output, with
 * no clock, no randomness and no I/O. That is what lets the Python simulator in
 * research/simulator reproduce production prices from the shared test vectors.
 */
export function runTick(input: TickInput): TickOutput {
  validateTickInput(input);

  const { tick, params, products, paidUnits } = input;
  const explain: Record<string, ExplainEntry> = {};
  const output: ProductState[] = [];

  for (const group of groupByGroupId(products)) {
    output.push(...runGroup(group, tick, params, paidUnits, explain));
  }

  // Preserve the caller's product order, which keeps snapshots stable.
  const byId = new Map(output.map((product) => [product.productId, product]));
  return {
    tick,
    products: products.map((product) => byId.get(product.productId) ?? product),
    explain,
  };
}

function groupByGroupId(products: readonly ProductState[]): ProductState[][] {
  const groups = new Map<string, ProductState[]>();
  for (const product of products) {
    const bucket = groups.get(product.groupId);
    if (bucket) {
      bucket.push(product);
    } else {
      groups.set(product.groupId, [product]);
    }
  }
  return [...groups.values()];
}

function runGroup(
  group: readonly ProductState[],
  tick: number,
  params: TickInput['params'],
  paidUnits: Readonly<Record<string, number>>,
  explain: Record<string, ExplainEntry>,
): ProductState[] {
  const result: ProductState[] = [];
  const working: Working[] = [];

  // Sold-out products leave the calculation and keep their last price (spec 4.2).
  for (const product of group) {
    if (!product.active) {
      explain[product.productId] = { ...NEUTRAL_EXPLAIN };
      result.push({ ...product });
    }
  }

  const activeProducts = group.filter((product) => product.active);
  if (activeProducts.length === 0) {
    return result;
  }

  // Steps 1-2: smoothed demand. Every active product counts, including the ones
  // an admin has pinned, so their history stays usable once the pin expires.
  for (const product of activeProducts) {
    const units = paidUnits[product.productId] ?? 0;
    const demand = params.ewmaLambda * units + (1 - params.ewmaLambda) * product.smoothedDemand;
    const previous = previousRawPrice(product);
    working.push({
      product,
      demand,
      previous,
      candidate: previous,
      explain: { ...NEUTRAL_EXPLAIN },
      pinned: overrideAppliesAt(product, tick),
    });
  }

  const expectedShares = expectedSharesOf(activeProducts);
  const totalDemand = working.reduce((sum, entry) => sum + entry.demand, 0);
  const groupIsIdle = totalDemand < DEMAND_EPSILON;

  // Steps 3-9: per-product signals and the candidate price.
  for (const entry of working) {
    const { product } = entry;
    const expected = expectedShares.get(product.productId) ?? 0;
    const observed = groupIsIdle ? expected : entry.demand / totalDemand;

    const x = expected > 0 ? clip((observed - expected) / expected, -1, 1) : 0;
    const r = (entry.previous - product.basePrice) / product.basePrice;
    const stockRatio = product.stockInitial > 0 ? product.stockAvailable / product.stockInitial : 0;
    const k =
      params.lowStockThreshold > 0
        ? Math.max(0, params.lowStockThreshold - stockRatio) / params.lowStockThreshold
        : 0;

    if (entry.pinned) {
      entry.explain = { x, r, k, delta: 0, c: 1 };
      entry.candidate = product.manualOverride?.price ?? product.currentPrice;
      continue;
    }

    const delta = clip(
      params.demandSensitivity * x - params.meanReversion * r + params.stockPressure * k,
      -params.maxStepPct,
      params.maxStepPct,
    );

    entry.explain = { x, r, k, delta, c: 1 };
    entry.candidate = entry.previous * (1 + delta);
  }

  // Step 10: renormalisation keeps the group's weighted average near the base
  // average, so a surge in one drink pushes its substitutes down.
  const free = working.filter((entry) => !entry.pinned);
  if (params.renormalize && free.length > 0) {
    renormalizeGroup(free);
  }

  // Steps 11-12: re-apply the per-tick cap, the hard limits and the rounding.
  for (const entry of working) {
    const { product } = entry;
    const capped = entry.pinned
      ? entry.candidate
      : clip(
          entry.candidate,
          entry.previous * (1 - params.maxStepPct),
          entry.previous * (1 + params.maxStepPct),
        );
    const raw = clip(capped, product.minPrice, product.maxPrice);

    explain[product.productId] = entry.explain;
    result.push(
      nextState(
        product,
        roundIntoRange(raw, product.minPrice, product.maxPrice, params.roundingCents),
        raw,
        entry.demand,
        tick,
      ),
    );
  }

  return result;
}

function nextState(
  product: ProductState,
  price: number,
  rawPrice: number,
  demand: number,
  tick: number,
): ProductState {
  const next: ProductState = {
    ...product,
    currentPrice: price,
    rawPrice,
    smoothedDemand: demand,
  };
  if (product.manualOverride && !overrideAppliesAt(product, tick)) {
    delete next.manualOverride;
  }
  return next;
}

/**
 * Expected shares default to an even split and are normalised to sum to 1, so a
 * partially configured group still produces a usable demand signal.
 */
function expectedSharesOf(activeProducts: readonly ProductState[]): Map<string, number> {
  const evenShare = 1 / activeProducts.length;
  const raw = activeProducts.map((product) =>
    product.expectedShare !== undefined && product.expectedShare > 0
      ? product.expectedShare
      : evenShare,
  );
  const total = raw.reduce((sum, value) => sum + value, 0);

  return new Map(
    activeProducts.map((product, index) => [
      product.productId,
      total > 0 ? (raw[index] ?? 0) / total : 0,
    ]),
  );
}

function renormalizeGroup(free: Working[]): void {
  for (let iteration = 0; iteration < RENORMALIZE_MAX_ITERATIONS; iteration += 1) {
    let numerator = 0;
    let denominator = 0;

    for (const entry of free) {
      const weight = entry.demand + RENORMALIZE_EPSILON;
      numerator += weight * entry.product.basePrice;
      denominator += weight * entry.candidate;
    }

    if (denominator <= 0) {
      return;
    }

    const c = numerator / denominator;
    for (const entry of free) {
      entry.candidate = clip(c * entry.candidate, entry.product.minPrice, entry.product.maxPrice);
      // Cumulative across iterations, so `explain` reports the total correction.
      entry.explain.c *= c;
    }

    if (Math.abs(c - 1) < RENORMALIZE_TOLERANCE) {
      return;
    }
  }
}

File: research/simulator/src/bolsa_sim/engine.py
"""Python port of packages/pricing-engine/src/engine.ts (spec 5.3).

This is the research implementation. It exists so the simulator can explore
parameter space without a Node process in the loop, and it is checked against
the production engine on every shared test vector (spec 5.6) by
tests/test_parity.py. The operations are kept in the same order as the
TypeScript version so that both produce bit-identical floats.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from typing import Mapping, Sequence

from .money import clip, round_into_range
from .params import DEMAND_EPSILON, RENORMALIZE_EPSILON, RENORMALIZE_MAX_ITERATIONS
from .params import RENORMALIZE_TOLERANCE, EngineParams


@dataclass(frozen=True)
class ManualOverride:
    price: int
    until_tick: int


@dataclass(frozen=True)
class ProductState:
    product_id: str
    group_id: str
    base_price: int
    min_price: int
    max_price: int
    current_price: int
    smoothed_demand: float = 0.0
    raw_price: float | None = None
    expected_share: float | None = None
    stock_initial: int = 0
    stock_available: int = 0
    active: bool = True
    manual_override: ManualOverride | None = None


@dataclass(frozen=True)
class ExplainEntry:
    x: float = 0.0
    r: float = 0.0
    k: float = 0.0
    delta: float = 0.0
    c: float = 1.0


@dataclass
class TickOutput:
    tick: int
    products: list[ProductState]
    explain: dict[str, ExplainEntry] = field(default_factory=dict)


@dataclass
class _Working:
    product: ProductState
    demand: float
    previous: float
    candidate: float
    x: float = 0.0
    r: float = 0.0
    k: float = 0.0
    delta: float = 0.0
    c: float = 1.0
    pinned: bool = False

    def explain(self) -> ExplainEntry:
        return ExplainEntry(x=self.x, r=self.r, k=self.k, delta=self.delta, c=self.c)


def _override_applies_at(product: ProductState, tick: int) -> bool:
    return product.manual_override is not None and tick <= product.manual_override.until_tick


def _previous_raw_price(product: ProductState) -> float:
    raw = product.raw_price if product.raw_price is not None else product.current_price
    return clip(raw, product.min_price, product.max_price)


def _expected_shares(active: Sequence[ProductState]) -> dict[str, float]:
    even = 1 / len(active)
    raw = [
        p.expected_share if p.expected_share is not None and p.expected_share > 0 else even
        for p in active
    ]
    total = sum(raw)
    return {
        p.product_id: (value / total if total > 0 else 0.0)
        for p, value in zip(active, raw, strict=True)
    }


def _renormalize(free: list[_Working]) -> None:
    for _ in range(RENORMALIZE_MAX_ITERATIONS):
        numerator = 0.0
        denominator = 0.0
        for entry in free:
            weight = entry.demand + RENORMALIZE_EPSILON
            numerator += weight * entry.product.base_price
            denominator += weight * entry.candidate

        if denominator <= 0:
            return

        c = numerator / denominator
        for entry in free:
            entry.candidate = clip(
                c * entry.candidate, entry.product.min_price, entry.product.max_price
            )
            entry.c *= c

        if abs(c - 1) < RENORMALIZE_TOLERANCE:
            return


def _run_group(
    group: Sequence[ProductState],
    tick: int,
    params: EngineParams,
    paid_units: Mapping[str, int],
    explain: dict[str, ExplainEntry],
) -> list[ProductState]:
    result: list[ProductState] = []

    for product in group:
        if not product.active:
            explain[product.product_id] = ExplainEntry()
            result.append(product)

    active = [product for product in group if product.active]
    if not active:
        return result

    working: list[_Working] = []
    for product in active:
        units = paid_units.get(product.product_id, 0)
        demand = params.ewma_lambda * units + (1 - params.ewma_lambda) * product.smoothed_demand
        previous = _previous_raw_price(product)
        working.append(
            _Working(
                product=product,
                demand=demand,
                previous=previous,
                candidate=previous,
                pinned=_override_applies_at(product, tick),
            )
        )

    expected_shares = _expected_shares(active)
    total_demand = sum(entry.demand for entry in working)
    group_is_idle = total_demand < DEMAND_EPSILON

    for entry in working:
        product = entry.product
        expected = expected_shares.get(product.product_id, 0.0)
        observed = expected if group_is_idle else entry.demand / total_demand

        entry.x = clip((observed - expected) / expected, -1, 1) if expected > 0 else 0.0
        entry.r = (entry.previous - product.base_price) / product.base_price
        stock_ratio = (
            product.stock_available / product.stock_initial if product.stock_initial > 0 else 0.0
        )
        entry.k = (
            max(0.0, params.low_stock_threshold - stock_ratio) / params.low_stock_threshold
            if params.low_stock_threshold > 0
            else 0.0
        )

        if entry.pinned:
            assert product.manual_override is not None
            entry.delta = 0.0
            entry.candidate = product.manual_override.price
            continue

        entry.delta = clip(
            params.demand_sensitivity * entry.x
            - params.mean_reversion * entry.r
            + params.stock_pressure * entry.k,
            -params.max_step_pct,
            params.max_step_pct,
        )
        entry.candidate = entry.previous * (1 + entry.delta)

    free = [entry for entry in working if not entry.pinned]
    if params.renormalize and free:
        _renormalize(free)

    for entry in working:
        product = entry.product
        capped = (
            entry.candidate
            if entry.pinned
            else clip(
                entry.candidate,
                entry.previous * (1 - params.max_step_pct),
                entry.previous * (1 + params.max_step_pct),
            )
        )
        raw = clip(capped, product.min_price, product.max_price)
        price = round_into_range(raw, product.min_price, product.max_price, params.rounding_cents)

        explain[product.product_id] = entry.explain()
        next_override = (
            product.manual_override if _override_applies_at(product, tick) else None
        )
        result.append(
            replace(
                product,
                current_price=price,
                raw_price=raw,
                smoothed_demand=entry.demand,
                manual_override=next_override,
            )
        )

    return result


def run_tick(
    tick: int,
    params: EngineParams,
    products: Sequence[ProductState],
    paid_units: Mapping[str, int] | None = None,
) -> TickOutput:
    paid_units = paid_units or {}
    explain: dict[str, ExplainEntry] = {}

    groups: dict[str, list[ProductState]] = {}
    for product in products:
        groups.setdefault(product.group_id, []).append(product)

    updated: dict[str, ProductState] = {}
    for group in groups.values():
        for product in _run_group(group, tick, params, paid_units, explain):
            updated[product.product_id] = product

    return TickOutput(
        tick=tick,
        products=[updated.get(p.product_id, p) for p in products],
        explain=explain,
    )
